<table> bordercolor attribute should not create a visible border

karlcow opened the pull request:
#### 9311f46770bac895da2a29aedd976598f44e1ca0
<pre>
&lt;table&gt; bordercolor attribute should not create a visible border
<a href="https://bugs.webkit.org/show_bug.cgi?id=256850">https://bugs.webkit.org/show_bug.cgi?id=256850</a>
rdar://109436009

Reviewed by Tim Nguyen.

A single bordercolor attribute on a table should not affect the
visibility of the border, which is none by default. This is aligned
with Firefox and what the HTML spec is prescribing.
&gt; treat the [bordercolor] attribute as a presentational hint setting
&gt; the element&apos;s &apos;border-top-color&apos;, &apos;border-right-color&apos;,
&gt; &apos;border-bottom-color&apos;, and &apos;border-left-color&apos; properties to the
&gt; resulting color.&quot;
&gt; — <a href="https://html.spec.whatwg.org/multipage/rendering.html#tables-2">https://html.spec.whatwg.org/multipage/rendering.html#tables-2</a>

It imports also one knew WPT test
WPT: 3d9cd84

There is a followup bug for tests failing
<a href="https://bugs.webkit.org/show_bug.cgi?id=257254">https://bugs.webkit.org/show_bug.cgi?id=257254</a>
This is the next step.

* LayoutTests/fast/table/border-changes-expected.txt:
* LayoutTests/fast/table/border-changes.html:
* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/resources/resource-files.json:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/colgroup_valign_bottom-expected.xhtml: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/colgroup_valign_top-expected.xhtml: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/table-bordercolor-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/table-bordercolor-001-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/table-bordercolor-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/w3c-import.log:
* LayoutTests/platform/glib/tables/mozilla/bugs/bug1163-1-expected.txt:
* LayoutTests/platform/glib/tables/mozilla/bugs/bug131020_iframe-expected.txt:
* LayoutTests/platform/glib/tables/mozilla/bugs/bug131020-expected.txt:
* LayoutTests/platform/gtk/tables/mozilla/bugs/bug1163-1-expected.png: Removed.
* LayoutTests/platform/ios-wk2/tables/mozilla/bugs/bug1163-1-expected.txt: Added.
* LayoutTests/platform/ios-wk2/tables/mozilla/bugs/bug131020_iframe-expected.txt: Added.
* LayoutTests/platform/ios-wk2/tables/mozilla/bugs/bug131020-expected.txt: Added.
* LayoutTests/platform/mac/tables/mozilla/bugs/bug1163-1-expected.png: Removed.
* LayoutTests/platform/mac/tables/mozilla/bugs/bug131020_iframe-expected.txt: Removed.
* LayoutTests/platform/mac/tables/mozilla/bugs/bug131020-expected.txt: Removed.
* LayoutTests/platform/wincairo/tables/mozilla/bugs/bug1163-1-expected.txt: Removed.
* LayoutTests/platform/wincairo/tables/mozilla/bugs/bug131020_iframe-expected.txt: Removed.
* LayoutTests/platform/wincairo/tables/mozilla/bugs/bug131020-expected.txt: Removed.
* LayoutTests/platform/wpe/tables/mozilla/bugs/bug1163-1-expected.txt: Removed.
* LayoutTests/platform/wpe/tables/mozilla/bugs/bug131020_iframe-expected.txt: Removed.
* LayoutTests/platform/wpe/tables/mozilla/bugs/bug131020-expected.txt: Removed.
* LayoutTests/tables/mozilla/bugs/bug1163-1-expected.txt:
* LayoutTests/tables/mozilla/bugs/bug131020_iframe-expected.txt:
* LayoutTests/tables/mozilla/bugs/bug131020-expected.txt:
* LayoutTests/TestExpectations:
* Source/WebCore/html/HTMLTableElement.cpp:
(WebCore::HTMLTableElement::additionalPresentationalHintStyle):
(WebCore::HTMLTableElement::attributeChanged):
(WebCore::HTMLTableElement::cellBorders const):
* Source/WebCore/html/HTMLTableElement.h:

Canonical link: <a href="https://commits.webkit.org/265670@main">https://commits.webkit.org/265670@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5bec4e7e1e19a98890524e8ba3bd46a5fd43c779

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11553 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11757 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12118 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13198 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11013 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11574 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14143 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11737 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13888 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11717 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12580 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9798 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13620 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9872 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10493 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17633 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10944 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10647 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13827 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11038 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9099 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10215 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/10372 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2783 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14492 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10895 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/fast/table/border-changes-expected.txt:
```diff
@@ -67,10 +67,10 @@ PASS getComputedStyle(cell, '').borderLeftWidth is '1px'
+FAIL getComputedStyle(cell, '').borderLeftStyle should be inset. Was solid.
+FAIL getComputedStyle(cell, '').borderRightStyle should be inset. Was solid.
+FAIL getComputedStyle(cell, '').borderTopStyle should be inset. Was solid.
+FAIL getComputedStyle(cell, '').borderBottomStyle should be inset. Was solid.
```
FAIL messages are not OK, we should change the test if needed.

karlcow replied:
Thanks @nt1m 
The test is good but WebKit fails for now.

This is STP 170 and Firefox side by side.
![Screenshot 2023-05-25 at 12 25 12](https://github.com/WebKit/WebKit/assets/505230/3759dc9e-b358-42f8-9dd6-65aae9cc76f8)

The Patch fixes some of the fails related to the report of Daniel Holbert. I discovered more secondary issue. 
And it's why I opened the followup bug https://bugs.webkit.org/show_bug.cgi?id=257254

What would you recommend?

nt1m replied:
Ah ok, I guess we can keep it failing and fix it later. Though ideally, we'd change and fix the test at the same time.

nt1m commented on LayoutTests/fast/table/border-changes-expected.txt:
```diff
@@ -103,39 +103,40 @@ PASS getComputedStyle(table, '').borderLeftWidth is '1px'
-PASS getComputedStyle(cell, '').borderBottomWidth is '1px'
+FAIL getComputedStyle(cell, '').borderTopWidth should be 0px. Was 1px.
+FAIL getComputedStyle(cell, '').borderBottomWidth should be 0px. Was 1px.
 PASS getComputedStyle(cell, '').borderLeftStyle is 'none'
 PASS getComputedStyle(cell, '').borderRightStyle is 'none'
-PASS getComputedStyle(cell, '').borderTopStyle is 'solid'
-PASS getComputedStyle(cell, '').borderBottomStyle is 'solid'
+FAIL getComputedStyle(cell, '').borderTopStyle should be none. Was solid.
+FAIL getComputedStyle(cell, '').borderBottomStyle should be none. Was solid.
```
ditto

karlcow replied:
see discussion in https://github.com/WebKit/WebKit/pull/14060#discussion_r1204954301

nt1m commented on LayoutTests/imported/w3c/resources/import-expectations.json:
```diff
@@ -304,6 +304,9 @@
     "web-platform-tests/html/rendering": "skip",
+    "web-platform-tests/html/rendering/non-replaced-elements/tables": "import",
+    "web-platform-tests/html/rendering/non-replaced-elements/tables/table-bordercolor-001-ref.html": "import",
+    "web-platform-tests/html/rendering/non-replaced-elements/tables/table-bordercolor-001.html": "import",
```
can you avoid adding unnecessary paths?

karlcow replied:
Could you explain? :) I'm not sure what is the goal of this file here. 
This was created probably when I reset the tests.

nt1m replied:
```suggestion
    "web-platform-tests/html/rendering/non-replaced-elements/tables": "import",
```
This is what I'm suggesting, the 2 other paths are included in the first one.

nt1m commented on LayoutTests/platform/mac-wk1/tables/mozilla/bugs/bug1163-1-expected.txt:
Is there a reason we now have WebKitLegacy specific results? Results should be consistent between wk2 and wk1 for this feature at least. 

You might want to use the platform/mac/ directory if you want to place results common to both.

karlcow replied:
I started with very generic tests but then it failed a couple of times. 
Hence why I started to add in specific platforms. Let's see. 

I can try again. I thought I had tried.

```
git diff ../../LayoutTests/tables/mozilla/bugs/bug1163-1-expected.txt ../../LayoutTests/platform/mac-wk1/tables/mozilla/bugs/bug1163-1-expected.txt
git diff ../../LayoutTests/platform/mac-wk2/tables/mozilla/bugs/bug1163-1-expected.txt ../../LayoutTests/platform/mac-wk1/tables/mozilla/bugs/bug1163-1-expected.txt
git diff ../../LayoutTests/platform/mac-wk2/tables/mozilla/bugs/bug1163-1-expected.txt ../../LayoutTests/platform/mac-ventura/tables/mozilla/bugs/bug1163-1-expected.txt
```

no differences.  😞 

@nt1m `LayoutTests/tables/mozilla/bugs/bug1163-1-expected.txt` should be enough? no? without having to rely on `/platform/mac/`

Thanks.

nt1m replied:
Yeah, in your case if the 3 files are identical, then `LayoutTests/tables/mozilla/bugs/bug1163-1-expected.txt` is enough.

karlcow replied:
@nt1m rha… so I removed them and now it fails… go figure

![Screenshot 2023-06-27 at 19 51 09](https://github.com/WebKit/WebKit/assets/505230/a05ca166-6a1b-4386-88bc-b2a9ab512ba2)

Ahmad-S792 replied:
@karlcow - Could it be because of 'png' expectation file?

karlcow replied:
Hmmm possible…  There are two of them.

```
LayoutTests/platform/mac/tables/mozilla/bugs/bug1163-1-expected.png
LayoutTests/platform/gtk/tables/mozilla/bugs/bug1163-1-expected.png
```

nt1m replied:
The PNG expectations don't do anything.

nt1m replied:
@karlcow I still see platform specific `.txt` expectations in LayoutTests/platform/mac/tables/mozilla/bugs/

cdumez commented on LayoutTests/platform/glib/tables/bugs/bug1163-1-expected.txt:
```diff
@@ -6,7 +6,7 @@ layer at (0,0) size 800x600
       RenderBlock (anonymous) at (0,0) size 784x18
         RenderText {#text} at (0,0) size 4x17
           text run at (0,0) width 4: " "
-      RenderTable {TABLE} at (0,18) size 210x92 [border: (3px solid #0000BB)]
```
You glib test expectations seem to be in the wrong folder, you dropped `mozilla` from the path for some reason.

nt1m reviewed (approved):
Can you update the commit title to say this is about the `bordercolor` attribute? (there isn't a `tableborder` attribute)

As discussed on Slack, `m_borderColorAttr` can also be removed entirely as well.

karlcow committed:
<table> bordercolor attribute should not create a visible border
https://bugs.webkit.org/show_bug.cgi?id=256850
rdar://109436009

Reviewed by Tim Nguyen.

A single bordercolor attribute on a table should not affect the
visibility of the border, which is none by default. This is aligned
with Firefox and what the HTML spec is prescribing.
> treat the [bordercolor] attribute as a presentational hint setting
> the element's 'border-top-color', 'border-right-color',
> 'border-bottom-color', and 'border-left-color' properties to the
> resulting color."
> — https://html.spec.whatwg.org/multipage/rendering.html#tables-2

It imports also one knew WPT test
WPT: 3d9cd84

There is a followup bug for tests failing
https://bugs.webkit.org/show_bug.cgi?id=257254
This is the next step.

* LayoutTests/fast/table/border-changes-expected.txt:
* LayoutTests/fast/table/border-changes.html:
* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/resources/resource-files.json:
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/colgroup_valign_bottom-expected.xhtml: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/colgroup_valign_top-expected.xhtml: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/table-bordercolor-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/table-bordercolor-001-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/table-bordercolor-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/w3c-import.log:
* LayoutTests/platform/glib/tables/mozilla/bugs/bug1163-1-expected.txt:
* LayoutTests/platform/glib/tables/mozilla/bugs/bug131020_iframe-expected.txt:
* LayoutTests/platform/glib/tables/mozilla/bugs/bug131020-expected.txt:
* LayoutTests/platform/gtk/tables/mozilla/bugs/bug1163-1-expected.png: Removed.
* LayoutTests/platform/ios-wk2/tables/mozilla/bugs/bug1163-1-expected.txt: Added.
* LayoutTests/platform/ios-wk2/tables/mozilla/bugs/bug131020_iframe-expected.txt: Added.
* LayoutTests/platform/ios-wk2/tables/mozilla/bugs/bug131020-expected.txt: Added.
* LayoutTests/platform/mac/tables/mozilla/bugs/bug1163-1-expected.png: Removed.
* LayoutTests/platform/mac/tables/mozilla/bugs/bug131020_iframe-expected.txt: Removed.
* LayoutTests/platform/mac/tables/mozilla/bugs/bug131020-expected.txt: Removed.
* LayoutTests/platform/wincairo/tables/mozilla/bugs/bug1163-1-expected.txt: Removed.
* LayoutTests/platform/wincairo/tables/mozilla/bugs/bug131020_iframe-expected.txt: Removed.
* LayoutTests/platform/wincairo/tables/mozilla/bugs/bug131020-expected.txt: Removed.
* LayoutTests/platform/wpe/tables/mozilla/bugs/bug1163-1-expected.txt: Removed.
* LayoutTests/platform/wpe/tables/mozilla/bugs/bug131020_iframe-expected.txt: Removed.
* LayoutTests/platform/wpe/tables/mozilla/bugs/bug131020-expected.txt: Removed.
* LayoutTests/tables/mozilla/bugs/bug1163-1-expected.txt:
* LayoutTests/tables/mozilla/bugs/bug131020_iframe-expected.txt:
* LayoutTests/tables/mozilla/bugs/bug131020-expected.txt:
* LayoutTests/TestExpectations:
* Source/WebCore/html/HTMLTableElement.cpp:
(WebCore::HTMLTableElement::additionalPresentationalHintStyle):
(WebCore::HTMLTableElement::attributeChanged):
(WebCore::HTMLTableElement::cellBorders const):
* Source/WebCore/html/HTMLTableElement.h:

Canonical link: https://commits.webkit.org/265670@main
